### DISCUSSION TITLE
ASK feature 개발

### DISCUSSION THREAD

**Discussion by kmsaidev** [euphony-io/euphony#85] [Knowledge Share]
ASK feature에 관해 같이 공부해보고 개발하기 위해 자유롭게 상의하고자 디스커션 남깁니다:)

좋은 자료 있으면 서로 공유하고 같이 이야기 나누면 좋을 것 같습니다!!
ASK를 이해하는 데 좋은 짧은 [동영상](https://youtu.be/eJ5m0Sbr2qw)이 있어 함께 공유합니다!

**Comment by psh2849** [comment]
감사합니다! 👍 추가로 조사해서 공유하도록 하겠습니다!

**Comment by kuro11pow2** [comment]
자료 감사합니다~ 한눈에 비교하는 이미지를 찾아서 첨부합니다
![image](https://user-images.githubusercontent.com/47289893/135952290-ee0ada82-8d1c-4aa0-a6a4-4399f506fcd5.png)

**Reply by HACore** [reply]
깔끔하게 보기 좋네요!👍

**Comment by judemin** [comment]
자료 공유 감사드립니다!!

**Comment by aiclaudev** [comment]
감사합니다!!

**Comment by jopopcorn** [comment]
코드를 읽어보면서 모르는 부분을 간단하게 정리해보았습니다!
찾아보고 이해한 대로 적어서 의미가 다를 수도 있는데, 정정할 부분이 있다면 말씀해주세요!

- **Modem(MOdulator and DEModulator)**
  - 의미: 신호를 변・복조 하는 장치. 디지털 신호를 아날로그로 변조하고(modulate), 아날로그 신호를 디지털 신호로 복조(demodulate)한다.
  - 파일 경로: `euphony/src/main/cpp/core/Modem.h`, `euphony/src/main/cpp/core/ModemFactory.h`
- **ASK (Amplitude Shift Keying)**
  - 의미: 디지털 신호 전송에 사용하는 변조 방식 중 하나로, 진폭(높낮이)으로 신호를 구분할 수 있다. 모뎀의 초기 방식이다.
  - 파일 경로: `euphony/src/main/java/co.euphony/util/EuOption.java`
- **FSK (Frequency Shift Keying)**
  - 의미: 디지털 변조 방식 중 하나로, 주파수로 신호를 구분한다.
  - 파일 경로: `euphony/src/main/cpp/core/FSK.h`, `euphony/src/main/cpp/core/source/FSK.cpp`

**Reply by aiclaudev** [reply]
정리 감사합니다! 해석해보고, 추가할 부분 등이 있으면 코멘트달겠습니다!

**Reply by Jsueeee** [reply]
깔끔하네요! 👍

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 

`euphony/src/main/cpp/core/Definitions.h` 파일의 37라인을 읽어보시면 0.7.1.6 버전에 ASK 피처가 있다는 주석이 적혀있습니다.
`v0.7.1.6` 브랜치에서 이전 버전 코드를 확인할 수 있는데, 내부 코드만 잘 살펴보면 피처 개발에 많은 도움이 될 것 같습니다.

modulate 함수 구현은 [`euphony/euphony/src/main/java/euphony/lib/transmitter/EuCodeMaker.java`](https://github.com/euphony-io/euphony/blob/v0.7.1.6/euphony/src/main/java/euphony/lib/transmitter/EuCodeMaker.java)의 코드를 참고하면 될 것 같습니다. 
0.8.0 버전의 [FSK 코드와 비슷한 흐름](https://github.com/euphony-io/euphony/discussions/101)인걸로 보이는데 다른 분들 생각은 어떤지 말씀해주세요!!
demodulate는 코드를 더 읽어봐야 감이 올 것 같습니다 ㅜㅜ

공부하면서 사소하게라도 알게 된 내용이 있다면 코멘트 올려주세요!!

**Reply by jopopcorn** [reply]
@euphony-io/euphoria 

단톡방에도 올렸지만 여기에서 좀 더 기록을 남겼으면 좋겠어서 다시 한번 말씀 드립니다.
저는 c++을 모르고 배경 지식부터 우선 이해해야겠다는 생각에 ASK, FSK 개념과 동향보고서 토대로 코드를 하나하나씩 분석해보았습니다.
하지만 확실하게 이해되는 부분이 거의 없어서 피처 개발을 어떻게 시작해야 할지 고민이 많았습니다.
이런 부분을 좀 더 자주 이야기를 나눴으면 좋았을텐데, 팀원분들 모두가 스케줄이 너무 바빴어서 그러지 못한 점이 아쉽습니다.
다른 분들은 어디까지 이해를 하셨고, 어떻게 개발을 시작하면 좋을지 말씀해주시면 좋을 것 같습니다!

우선 성호님께서도 c++을 몰라서 코드 분석부터 어려웠고 막막했다는 의견이 있었습니다.
저도 이부분을 공감했고, 그래서 c++을 알고 계시는 나머지 분들의 생각이 궁금합니다.

그리고 유닛테스트로 먼저 접근해서 개발을 시작해보자는 성호님의 의견이 있었는데, 이부분은 어떻게 생각하시는지 말씀해주심 감사하겠습니다!

**Reply by designe** [reply]
솔직하고 좋은 의견 개진 감사드립니다 :)
Euphony 라이브러리에서 쓰이는 몇몇 cpp에서 알아두면 좋은 개념이 있다면

1. STL : STL에서도 vector 개념이 많이 사용됩니다. 이건 java나 kotlin 에서의 동적 배열과 동일합니다.
2. 스마트포인터 : 간혹 shared_ptr, unique_ptr 개념도 등장합니다. cpp에서 포인터를 통한 메모리 관리가 정말 어렵거든요. 스마트포인터를 쓰면 알아서 관리 해주기 때문에 개발자는 알고리즘 개발에 집중할 수 있는 점이 있습니다.
3. gtest : 아마도 gtest 또한 어렵고 무섭게 느껴지는 부분일 수 있을 것 같습니다. gtest는 junit이나 다른 유닛테스트와 거의 유사하긴 합니다. 이 곳에서도 parameterized test를 사용하고 있는데 다른 몇몇 유닛테스트를 참고해보면 금방 이해하실 수 있을거에요 :)

(위 세가지에 대해 한번 디테일하게 조사해 discussion에 지식 공유해주실 분 있을까요? ㅎㅎ)

성호님이 말씀해주신대로 유닛테스트부터 시작해 개발하는 방법이 매우 좋은 방법이니 그렇게 시작해보는 것도 좋을 것 같아요.
한번 유닛테스트부터 ask는 어떻게 나오면 좋을지 한번 구성해보시면 좋을 것 같습니다.

@kuro11pow2 @YoungSeokHong @ycs1m1yk 
여기 3분이 함께 고민해주시면 좋겠습니다!

**Reply by psh2849** [reply]
저도 확실히 이해는 못했지만,
FSK같은 경우는 받아온 신호 값을 가지고 주파수를 변화시키는 방식입니다. 따라서,
```c++
result.push_back(
                        Wave::create()
                                .vibratesAt(kStandardFrequency + ((c - '0') * kFrequencyInterval))
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
```                
위의 코드에서 `kStandardFrequency`나 `kFrequencyInterval` 등을 이용해서 주파수를 변화시키는 것으로 보입니다.
하지만 ASK는 진폭만 변화시켜주면 되기 때문에 vibratesAt 값을 적절히 변경하면 되지 않을까..? 라는 생각이 드네요!
그리고 입력받는 케이스도 0과 1로 구분해서 구현하면 되지 않을까요?

**Reply by designe** [reply]
@psh2849 맞아요! 0 또는 1이기 때문에 BASE2를 활용하면 좋을 것 같아요 :)

**Reply by psh2849** [reply]
그럼 입력으로 들어온 값을 2진수로 변경시키고 2진수 값들을 하나씩 불러와서 0,1로 구분한 다음에 진폭 변경해주면 되지 않을까요???

**Reply by designe** [reply]
0은 amplitude가 없는 경우, 1은 amplitude가 발생하는 경우 1로 판단하면 될 것 같네요!

**Reply by HACore** [reply]
저도 유닛테스트로 시작하는게 첫 스타트로 좋아보여요!
`euphony/src/main/cpp/tests/FSKTest.cpp`의 구조를 참고하면 될것같은데 조금 더 파악해보고 글 작성해보겠습니다..!

**Reply by jopopcorn** [reply]
> 저도 유닛테스트로 시작하는게 첫 스타트로 좋아보여요! `euphony/src/main/cpp/tests/FSKTest.cpp`의 구조를 참고하면 될것같은데 조금 더 파악해보고 글 작성해보겠습니다..!

넵넵 좋습니다!

**Comment by hoho00** [comment]
## 스마트 포인터

잘 모르지만, 얕은 지식으로 떠듬떠듬 적어 봅니다. 추가할 내용 있으시면 추가 해 주세요.

c, c++ 은 포인터를 통해서 직접 memory에 access 합니다. 그렇기 때문에 그리고 malloc이나 new 연산자로 생성된 객체들은 해당 메모리를 직접 해제 해 주지 않으면 사용하지 않는 메모리가 남습니다.(memory leak) 그리고 현재 가르키고 있는 객체를 잘못 생성해 버리면, 포인터 객체가 가르키고 있는 객체가 없어지는 에러가 생기게 됩니다. (dangling pointer) 즉,

1. 메모리에 객체를 만들었는데 그걸 가르키고 있는 포인터가 없어서 사용할 수 없는 메모리 낭비가 생겼다 → memory leak
2. 해당 객체의 메모리를 두개 이상의 포인터가 가르키고 있다가, 가르키고 있는 객체가 해제되어서 남은 포인터가 없는 객체를 가르키고 있다. → dangling pointer 

위와 같은 메모리 사용 문제를 스마트 포인터를 사용하면 자동적으로 해결해 주기 때문에, raw pointer를 쓰는 것 보다 raw pointer를 스마트 포인터에 대입해서 사용합니다. 

euphony 라이브러리에서도 두가지의 스마트 포인터가 사용되더라구요.

```cpp
shared_pointer, unique_pointer 
```

스마트 포인터 덕분에 코드내에 따로 destruct 하는 부분도 보이지 않구요.

일단 unique포인터에 대해 알아보겠습니다.

### unique_pointer

간단하게 말씀드리자면, 하나의 객체에 하나의 포인터만 대입 할 수 있다 입니다. move를 통해 소유권을 이전할 수 있지만, 복사할 수 없습니다. memory leak를 예방 할 수 있겠네요. 왜냐하면 메모리 누수 같은 경우에는 하나의 객체에 여러가지 포인터를 참조하다가 하나의 포인터에서 해제 명령을 내리는 경우 발생 하는데, unique_pointer를 사용하면 하나의 객체에 하나의 포인터만 적용할 수 있기 때문에 memory leak가 발생하는 경우는 없겠네요. 

euphony 코드를 볼게요

```cpp
private:
	std::unique_ptr<FFTModel> fftModel
```

FFTModel의 경우에는 하나의 객체를 생성하면 여러 포인터가 가르킬 일 없으니까 unique포인터로 지정 해 둔 모습을 보입니다.

```cpp
fft = std::make_unique<FFTProcessor>(inputFFTSize, inputSampleRate);
```

test code 내에서 테스트 객체를 생성할 때도 make_unique함수를 사용해 unique 포인터를 생성하는 것을 알 수 있습니다. 유니크 포인터로 객체를 생성하면 매모리 해제 또한 자동으로 해 주기 때문에 그 이점때문에 스마트 포인터를 쓰려면 그냥 유니크 포인터로 객체를 생성하면 됩니다. 

### shared_pointer

이것은 하나의 객체에 여러개의 포인터를 가르킬 수 있지만, shared_pointer자체에서 객체에 가르키는 pointer의 수를 세기 때문에 객체를 가르키는 pointer의 수가 0가 되면 해당 객체를 해제 합니다. dangling pointer를 예방 할 수 있겠네요.

```cpp
std::shared_ptr<Packet> demodulate(const WaveList& waveList);
```

Packet의 벡터, 즉 HexVector 같은 경우 하나의 객체가 여러개의 포인터를 가르켜야 할 일이 있는 것 같습니다. 그래서 shared_pointer를 쓴 것 같습니다.

**Reply by designe** [reply]
@hoho00 좋네요! 이거 discussion 분기해서 공유해도 좋을 것 같습니다.

**Reply by hoho00** [reply]
넵!!

**Reply by psh2849** [reply]
감사합니다!

**Comment by jopopcorn** [comment]
@euphony-io/euphoria 
생각해본 ASK 변조 로직의 흐름인데 이렇게 구현하면 괜찮을까요? 더 좋은 방법이 있다면 말씀해주세요!
1. 입력 받은 문자열 `source`의 HexVector을 만든다. (`ASCIICharset().encode(source);`) 
2. HexVector로 변환한 값을 Base2 타입의 패킷으로 만든다. (`new Packet(BaseType::BASE2, hexVector);`)
3. WaveList FSK::modulate(Packet* packet) 함수에서 `return this->modulate(packet->toString());`를 수행하여 패킷을 문자열로 바꾸고 변조를 시작한다.
4. WaveList ASK::modulate(std::string code) 함수에서 'S', '0', '1'에 따른 아날로그 음파를 생성한다.

**Reply by jopopcorn** [reply]
0과 1로만 이루어졌대도 음파를 n번 혹은 무한대로 생성하려면 패킷이 필요할 것 같아서 이런 흐름으로 생각해봤습니다!
만약 패킷이 없어도 괜찮다면 어떤식으로 구별할 수 있는지도 궁금하네요..

그리고 16진수에서 2진수로 변환하면 0과 1로만 이루어졌기 때문에 Base2Exception은 없어도 괜찮은걸까요??..
혼자 구현을 시도해보다가 궁금해서 추가로 질문(?) 남깁니다!

**Reply by psh2849** [reply]
넵 맞는 것 같습니다! 혹시 테스트 케이스는 생각해보셨나요??

**Reply by jopopcorn** [reply]
> 넵 맞는 것 같습니다! 혹시 테스트 케이스는 생각해보셨나요??

gtest에서 사용하는 메서드들이 뭘 의미하는지 찾아보고 FSK랑 다른 테스트코드 참고하느라 아직 제대로 작성하지는 않았습니다!

**Comment by jopopcorn** [comment]
유포니 라이브러리 내의 여러 gtest 코드들을 읽어보고 자주 보이는 명령어, 기능들을 간단하게 정리해보았습니다.
[gtest 레퍼런스](https://google.github.io/googletest/reference/testing.html)와 [블로그 포스팅](https://surpreem.com/%EA%B5%AC%EA%B8%80-c-%ED%85%8C%EC%8A%A4%ED%8A%B8-%ED%94%84%EB%A0%88%EC%9E%84%EC%9B%8C%ED%81%AC-%EA%B3%A0%EA%B8%89-%EC%A3%BC%EC%A0%9C/)을 참고하여 정리했으니 테스트 코드 작성하는데 참고하시면 좋을 것 같습니다!

XX 클래스마다 이름을 다르게 짓는 것을 생각해서 임의로 넣은 것이고, 명령어 뒤에 오는 X는 그 뒤에 들어오는 다양한 기능들이 있다는 것을 표시한 것입니다!
유포니 라이브러리의 gtest 파일을 열어보면서 확인하시면 더 이해가 잘될 것 같습니다!

</br>

class XXTestFixture : public ::testing::TestWithParam<TestParamType>
-> 값 매개변수 테스트를 만들기 위해 픽스처 클래스를 정의함.

**ASSERT_X**: 어설션이 실패할 경우 프로그램 실행을 중단함
**EXPECT_X**: 테스트 케이스에 실패해도 프로그램을 계속 실행함

**EXPECT_EQ**: 두 값이 같기를 기대하는 메서드
**ASSERT_TRUE**: 항상 true인 조건을 기대하는 메서드
**ASSERT_NE**: 두 값이 일치하지 않기를 기대하는 메서드

</br>

**TEST_P**: 테스트 픽스처 클래스에서 사용하는 테스트 패턴을 정의함. 테스트 픽스처 클래스를 의미하는 `TestName`과, 테스트 수트를 의미하는 `TestFixtureName`을 매개변수로 지정함. 이 안에서 ASSERT_X, EXPECT_X와 같은 명령문을 활용할 수 있음


```c++
TEST_P(TestFixtureName, TestName) {
  ... statements ...
}
```

</br>

**INSTANTIATE_TEST_SUITE_P**: 테스트 케이스인 TestParamType들을 선언하고 테스트를 확인하도록 함

**Reply by psh2849** [reply]
테스트 케이스를 어떻게 짜야될지 모르겠네요 ㅠㅠ
```c++
TestParamType("0", 0, 0),
```
앞엔 2진수 값이 들어가야 할 것 같고, 결과는 0 또는 1로 출력해줘야할까요..?

**Reply by jopopcorn** [reply]
> 테스트 케이스를 어떻게 짜야될지 모르겠네요 ㅠㅠ
> 
> ```c++
> TestParamType("0", 0, 0),
> ```
> 
> 앞엔 2진수 값이 들어가야 할 것 같고, 결과는 0 또는 1로 출력해줘야할까요..?

FSK에서는 입력받은 문자열, 기대하는 문자열 길이, 기대하는 주파수 인덱스가 들어가더라고요!
제 생각에는 ASK에서는 첫째 둘째 매개변수는 그대로 하고, 세번째 매개변수에는 기대하는 진폭(0,1) 아니면 입력받은 문자열에 대한 이진값들이 필요하지 않을까요..??

**Comment by ycs1m1yk** [comment]
FSK 테스트코드에서 테스트 패턴을 정의하는 TEST_P 부분을 살펴봅시다.
https://github.com/euphony-io/euphony/blob/c1bb8f4841ebf8e7cad03c2245d056ce8980b41a/euphony/src/main/cpp/tests/FSKTest.cpp#L32-L47

### 테스트 파라미터 형식과 작성
```cpp
std::tie(inputCode, expectedCodeLength, expectedFreqIndex) = GetParam();
```
위와 같이 파라미터를 받아서 테스트하는것을 알 수 있습니다.

따라서 테스트케이스도 아래와 같이 작성하면됩니다.
```cpp
TestParamtype(inputCode, expectedCodeLength, expectedFreqIndex)
```

### 무엇을 테스트하는가?
TEST_P 에서는 2개의 EXPECT_EQ 메서드로 테스트가 이루어집니다.

1.

```cpp
auto modulateFSK = fsk->modulate(inputCode);
    EXPECT_EQ(modulateFSK.size(), expectedCodeLength);
```

2.

```cpp
auto demodulateResult = fsk->demodulate(modulateFSK);
    EXPECT_EQ(inputCode, demodulateResult->getPayloadStr());
```

3번째 파라미터인 expectedFreqIndex 에 대한 테스트는 수행되지 않고 있습니다.

### 실제 테스트케이스 예시

https://github.com/euphony-io/euphony/blob/c1bb8f4841ebf8e7cad03c2245d056ce8980b41a/euphony/src/main/cpp/tests/FSKTest.cpp#L65

FSKTest.cpp의 첫번째 테스트 케이스입니다.
1. 첫번째 파라미터 `"0"`은 string타입의 inputCode이며 demodulateResult->getPayloadStr()과 비교됩니다.
2. 두번째 파라미터 `1` 은 expectedCodeLength이며 modulateFSK.size()와 비교됩니다.
   string "0"의 길이가 1이죠.
3. 세번째 파라미터 `0`은 expectedFreqIndex입니다. TEST_P에서 이에대한 테스트 메서드가 정의돼있지 않습니다. 실제로 이 값을 임의로 바꾸어서 빌드해도 gtest는 통과됩니다.

### ASK의 테스트케이스
```cpp
TestParamtype("0", 1, 0),
TestParamtype("1", 1, 1),
TestParamtype("01", 2, 0),
TestParamtype("1101", 4, 0),
TestParamtype("1010111000", 10, 0)
```

위와 같이 첫번째에는 0이나1 문자로 이루어진 inputCode, 두번째에는 inputCode의 길이를 넣어주면됩니다.  세번째 파라미터는 무시해도 좋습니다. 일단 임의의 값을 넣어놓으면 될것입니다.

혹은, TEST_P에서
https://github.com/euphony-io/euphony/blob/c1bb8f4841ebf8e7cad03c2245d056ce8980b41a/euphony/src/main/cpp/tests/FSKTest.cpp#L40

위 라인을 다음과 같이 수정합니다.
```cpp
std::tie(inputCode, expectedCodeLength) = GetParam();
```
그리고 테스트케이스를
```cpp
TestParamtype("0", 1),
TestParamtype("11", 2)
```
위와 같이 작성하면 될 것입니다.

**Reply by jopopcorn** [reply]
오호 위에서 테스트케이스 매개변수를 뭘로 넣을지 이야기를 나누고 있었는데 세번째 파라미터는 없어도 되는군요!
덕분에 확실히 이해하고 테스트코드 작성할 수 있겠네요 정말정말 감사합니다!! 👍👍

**Reply by psh2849** [reply]
늦은 시간까지 도와주셔서 감사합니다ㅜㅜ 사이즈 값만 넣어줘도 되는군요..

**Reply by ycs1m1yk** [reply]
저도 궁금해서 조사해본건데 도움되면 좋겠네요! 응원합니다 😄

**Comment by ycs1m1yk** [comment]
# Amplitude를 어떻게 설정할까요?

### FSK에서는...

https://github.com/euphony-io/euphony/blob/c1bb8f4841ebf8e7cad03c2245d056ce8980b41a/euphony/src/main/cpp/core/source/FSK.cpp#L19-L34

FSK클래스의 modulate 메서드의 일부분입니다.

파라미터 string code를 받아서
```cpp
for (char c : code )
```
code의 문자마다 char c로 for문이 돌아갑니다.

하나의 문자 c마다 하나의 Wave가 생성됩니다.

c의 값에 따라서 case문으로 분기하고 .vibratesAt() 메서드로 frequency값을 설정합니다. (Frequency Shift)

### Amplitude 속성은 어디에?

ASK도 FSK처럼 modulate 메서드안에서 문자에 따라서 amplitude를 설정하면 되겠죠?
 
https://github.com/euphony-io/euphony/blob/c1bb8f4841ebf8e7cad03c2245d056ce8980b41a/euphony/src/main/cpp/core/Wave.h#L40-L49

그런데 Wave.h를 살펴보면 진폭에 관한 속성이 정의돼있지 않습니다.

그냥 mAplitude 속성을 추가하면 되는걸까요?

#### 참고
![image](https://user-images.githubusercontent.com/27720475/136707558-cb20973a-dd1e-432c-b618-1dc908c343bc.png)

유포니 프로젝트내에서 amplitude 키워드를 검색해보면

EuPIOscillator와 Oscillator에서는 진폭을 다루고있네요.

근데 Wave클래스와는 연결점이 없는것 같습니다. ASK를 개발하는데 상관이 있는지 모르겠네요.

**Comment by kuro11pow2** [comment]
너무 길어져서 새로운 discussion으로 분리했습니다!

[FSK, Wave, Packet 그리고 ASK 짧은? 요약](https://github.com/euphony-io/euphony/discussions/115)
